File: FakturaCZ-Windows/app/services/recurring_scheduler.py
"""Recurring invoice scheduler — generates invoices from templates."""

from datetime import date, timedelta

from app.models.invoice import Invoice, InvoiceLine
from app.models.recurring import RecurringTemplate, LineItemSpec
from app.services import persistence


def process_templates() -> list[Invoice]:
    """Check all active templates and generate invoices for those due today."""
    templates = persistence.load_recurring()
    generated: list[Invoice] = []

    for template in templates:
        if not template.is_due_today:
            continue

        invoice = _generate_invoice(template)
        generated.append(invoice)
        template.advance()

    if generated:
        persistence.save_recurring(templates)

    return generated


def _generate_invoice(template: RecurringTemplate) -> Invoice:
    today = date.today()
    number = persistence.next_invoice_number()
    due = today + timedelta(days=template.payment_term_days)

    invoice = Invoice(
        invoice_number=number,
        type="STANDARD",
        state="OPEN",
        issued_date=today.isoformat(),
        taxable_supply_date=today.isoformat(),
        due_date=due.isoformat(),
        supply_code=template.supply_code,
        subject_registration_no=template.subject_registration_no,
        note=template.default_note,
    )

    lines = []
    for i, spec_dict in enumerate(template.line_items):
        spec = LineItemSpec.from_dict(spec_dict)
        desc = template.resolve_description(today) or spec.description
        line = InvoiceLine(
            item_description=desc,
            quantity=__import__("decimal").Decimal(spec.quantity),
            unit_price=__import__("decimal").Decimal(spec.unit_price),
            vat_rate=spec.vat_rate,
            unit=spec.unit,
            sort_order=i,
        )
        lines.append(line.to_dict())

    invoice.lines = lines
    persistence.upsert_invoice(invoice)
    return invoice

File: FakturaCZ/FakturaCZ/Services/RecurringScheduler.swift
import Foundation
import SwiftData

/// Cron-equivalent scheduler that checks recurring templates and generates invoices.
actor RecurringScheduler {
    static let shared = RecurringScheduler()

    private init() {}

    /// Checks all active recurring templates and generates invoices for those due today.
    /// Should be called on app launch and periodically via background tasks.
    @MainActor
    func processTemplates(context: ModelContext) async throws -> [Invoice] {
        let predicate = #Predicate<RecurringTemplate> { template in
            template.isActive
        }
        let descriptor = FetchDescriptor<RecurringTemplate>(predicate: predicate)
        let templates = try context.fetch(descriptor)

        var generated: [Invoice] = []

        for template in templates {
            guard template.isDueToday else { continue }

            let invoice = try generateInvoice(from: template, context: context)
            generated.append(invoice)

            template.advance()
        }

        if !generated.isEmpty {
            try context.save()
        }

        return generated
    }

    /// Creates a new Invoice from a RecurringTemplate.
    @MainActor
    private func generateInvoice(
        from template: RecurringTemplate,
        context: ModelContext
    ) throws -> Invoice {
        let invoiceNumber = try InvoiceNumberGenerator.next(context: context)
        let now = Date()
        let dueDate = Calendar.current.date(
            byAdding: .day,
            value: template.paymentTermDays,
            to: now
        )!

        let invoice = Invoice(
            invoiceNumber: invoiceNumber,
            type: .standard,
            issuedDate: now,
            taxableSupplyDate: now,
            dueDate: dueDate,
            supplyCode: template.supplyCode
        )
        invoice.subject = template.subject
        invoice.note = template.defaultNote

        // Decode line item specs and create InvoiceLine objects.
        if let lineData = template.lineItemsJSON {
            let decoder = JSONDecoder()
            if let specs = try? decoder.decode([LineItemSpec].self, from: lineData) {
                for (index, spec) in specs.enumerated() {
                    let resolvedDescription = template.resolveDescription(for: now)
                        .isEmpty ? spec.description : template.resolveDescription(for: now)

                    let line = InvoiceLine(
                        itemDescription: resolvedDescription,
                        quantity: spec.quantity,
                        unitPrice: spec.unitPrice,
                        vatRate: VATRate(rawValue: spec.vatRate) ?? .standard,
                        unit: spec.unit,
                        sortOrder: index
                    )
                    line.invoice = invoice
                    context.insert(line)
                }
            }
        }

        context.insert(invoice)
        return invoice
    }
}
